### ISSUE TITLE
Unexpected `store.sub` behavior for derived atoms.

### ISSUE THREAD

**Issue by elliotwaite** [pmndrs/jotai#1014]
Here's a [demo](https://codesandbox.io/s/jotai-createstore-derived-sub-j7qmrr?file=/src/App.tsx).

When subscribed to a derived atom, the number of times the callback is called seems to depend on whether the value for that derived atom is shown.
* If not shown, the callback is called once per update (expected).
* If shown, the callback is called twice per update (unexpected).
* When toggled from not shown to shown, the callback is called once (unexpected).

**Comment by dai-shi** [comment]
Thanks for reporting!
Hmm, I'm not sure how it's happening. Not even sure if it's expected implementation-wise, or is fixable.
Would you be interested in digging into it??

**Comment by elliotwaite** [comment]
I don't think I understand the Jotai internals well enough yet to debug this one, but I'll keep learning and I'll let you know if I make any progress on this.

**Comment by dai-shi** [comment]
No worries!
@Aslemammad can you work on this?

**Comment by elliotwaite** [comment]
So I think the reason the subscription callback gets called twice is that it gets called first when `derivedAtom` is invalidated and then again when its new value is read.

I tried to map out the steps through the code below, with each indent somewhat representing a new call into a function. The setup is that there is a `countAtom` and a `derivedAtom` that just returns the same value as `countAtom`, and we are subscribed to changes in `derivedAtom`. Then we start off the calls below by calling `setCount((c) => c + 1)` to increment the value of `countAtom` from 0 to 1.

Also, all the functions mentioned below are in `store.ts` unless specified otherwise, and I've left out all the `version` function arguments for simplicity since they were always `undefined` in this example.

```
// From the app code:
setCount(update: (c) => c + 1)

  // This calls the setAtom callback in useSetAtom (useSetAtom.ts) which then calls:
  store[WRITE_ATOM](writingAtom: countAtom, update: (c) => c + 1)

    // Which calls the writeAtom function in store.ts. Now from within writeAtom:
    writeAtomState(writingAtom: countAtom, update: (c) => c + 1)

      // From writeAtomState:
      invalidateDependents(atom: countAtom)

        // From invalidateDependents:
        setAtomInvalidated(atom: derivedAtom)

          // From setAtomInvalidated:
          setAtomState(atom: derivedAtom, atomState: {
            ... prevAtomState
            i: prevAtomState.r // Sets the invalidated revision.
          })

            // From setAtomState, add derivedAtom to the pendingMap:
            pendingMap.set(derivedAtom, prevAtomState)

    // From back in writeAtom (the original writeAtom for countAtom):
    flushPending()

      // From flushPending, we get derivedAtom from the pendingMap and call all
      // its listeners, which calls our subscription callback the first time.
      mounted?.l.forEach((listener) => listener())

        // If derivedAtom is also being read, this starts phase 2, where one of the
        // listeners that also gets called is the useAtomValue Reducer dispatch
        // callback (in useAtomValue.ts). This gets called to notify the Reducer
        // that the value of the atom has changed and it should read in the new
        // value. So the dispatch callback then calls its own getAtomValue function
        // (in useAtomValue.ts) which then calls:
        store[READ_ATOM](readingAtom: derivedAtom)
  
          // Which calls the readAtom function back in store.ts. Now from within readAtom:
          readAtomState(readingAtom: derivedAtom)
  
            // From readAtomState:
            setAtomPromiseOrValue(atom: derivedAtom, promiseOrValue: 1, dependencies: Set([countAtom]))
  
              // From setAtomPromiseOrValue:
              setAtomValue(atom: derivedAtom, promiseOrValue: 1, dependencies: Set([countAtom]))
  
                // From setAtomValue:
                setAtomState(atom: derivedAtom, atomState: {v: 1, r: 2, d: Map({countAtom: 2})
  
                  // From setAtomState, derivedAtom gets added to the pendingMap again:
                  pendingMap.set(derivedAtom, prevAtomState)

      // Now back in flushPending, since pendingMap.size is no longer zero, we go
      // back through the flushPending while loop and get derivedAtom from the
      // pendingMap and call all its listeners, which calls our subscription callback
      // the second time.
      mounted?.l.forEach((listener) => listener())
```

So in my original comment for this issue, I mentioned these three cases:
> * If not shown, the callback is called once per update (expected).
> * If shown, the callback is called twice per update (unexpected).
> * When toggled from not shown to shown, the callback is called once (unexpected).

* The first case is the same code sequence as above, except we skip phase 2 (callback is only called once).
* The second case is the exact code sequence as above (callback is called twice).
* And the third case, where `derivedAtom` toggles from not-shown to shown after having been changed when it was hidden, we get the same phase 2 as above, the only difference being that it is delayed until `derivedAtom` is read, and instead of `flushPending` being called from `writeAtom`, it's called from `commitAtom`, which gets called from a `useEffect` in `useAtomValue.ts` that gets called whenever the component reading `derivedAtom` re-renders.

In thinking about potential solutions, I noticed that all the unnecessary secondary calls were when the state was transitioning from an invalidated revision to a valid revision. So I tried changing the code in `flushPending` to only call the listeners if the state was not going from an invalidated revision to a valid revision.

In the `flushPending` function in `store.ts` I replaced:
```
const mounted = mountedMap.get(atom)
mounted?.l.forEach((listener) => listener())
```
... with:
```
if (
  !(
    prevAtomState &&
    'i' in prevAtomState &&
    atomState &&
    !('i' in atomState)
  )
) {
  const mounted = mountedMap.get(atom)
  mounted?.l.forEach((listener) => listener())
}
```

All the tests in the test suite passed with this change. Perhaps this means that it's okay to skip calling the listeners when the state is going from an invalid revision to a valid one because the listeners would have already been called when it was set to the invalid revision.

However, I wasn't sure if there were certain cases that may be negatively affected by this change, especially in regards to promises and errors. One way this change could potentially be made safer is to only skip calling the listeners if also both the previous state and current state were for values (not promises or errors), by also checking for `'v' in state`:

```
if (
  !(
    prevAtomState &&
    'i' in prevAtomState &&
    atomState &&
    !('i' in atomState) &&
    'v' in prevAtomState &&
    'v' in atomState
  )
) {
  const mounted = mountedMap.get(atom)
  mounted?.l.forEach((listener) => listener())
}
```

I'm also not sure how this change interplays with concurrent mode, or how it would affect performance, but if these listener calls are in fact unnecessary, it seems like it might be a good idea to skip them.

I was also wondering if there was a more elegant way of accomplishing this rather than just doing this check right before the listeners were called. Perhaps something could be changed earlier in the pipeline.

**Comment by dai-shi** [comment]
> the reason the subscription callback gets called twice is that it gets called first when derivedAtom is invalidated and then again when its new value is read.

Oh, I see...
If this is the case, it can be said working as expected.
You need the first callback anyway because the atom may not be used in components.

> All the tests in the test suite passed with this change.

That sounds weird. You might want to try with react 18. I also check current rendering behavior with https://github.com/dai-shi/will-this-react-global-state-work-in-concurrent-rendering.

**Comment by dai-shi** [comment]
```js
if (
  !(
    prevAtomState &&
    'i' in prevAtomState &&
    atomState &&
    !('i' in atomState)
  )
) {
```
Ah, I see, this is skipping the second one. Yeah, this may pass all tests.

> if these listener calls are in fact unnecessary

can't tell. we don't even know how the final concurrent rendering behaves. so, we shouldn't change it at this moment.

my gut feeling is it may not work with experimental "versioned write".

> Perhaps something could be changed earlier in the pipeline.

Yeah, if possible, that's should be the way to go. It might improve the performance.
Unless we have more tests with concurrent rendering behaviors, we can't be confident...

**Comment by elliotwaite** [comment]
> can't tell. we don't even know how the final concurrent rendering behaves. so, we shouldn't change it at this moment.
>
> my gut feeling is it may not work with experimental "versioned write".

Okay, sounds good. I'll wait till concurrent rending gets more stable before looking into this further. The extra sub calls should be somewhat harmless, but they may be surprising to users.

Also, I experimented with moving the logic to earlier in the pipeline by preventing those invalid->valid state transitions from being added to `pendingMap` in the first place, but that ended up breaking tests. I think this happens because when those state transitions aren't added to `pendingMap`, these lines in `flushPending` don't get called for them:
```
          if (atomState && atomState.d !== prevAtomState?.d) {
            mountDependencies(atom, atomState, prevAtomState?.d)
          }
``` 
... which as I understand it, updates the dependencies for the atom. And since those state transitions (invalid->valid) are when the atom is actually read, it's important to update the dependencies for the atom after that, because it's during the read when the dependencies are collected. So perhaps the logic was already at a good place. But yeah, I'll have to wait for a more stable React 18 to experiment with this further.

**Comment by dai-shi** [comment]
Sounds good. flushPending for invalid->valid is by design at the moment. If we find a better solution, it will be super nice, but I  expect we need a fairly big rewrite for it. As I said, we would need to add some tests for concurrent rendering before such a rewrite.

**Comment by TwistedMinda** [comment]
@dai-shi What's the state of React 18 concurrent mode stability and this issue in general?
Should you run some tests again? Or put "on hold" for now?

**Comment by dai-shi** [comment]
Concurrent rendering in react 18 should be stable.

Re: the state of this issue

I didn't closely look this issue yet, but:
- what's described in this issue seems valid,
- it's not high priority because it's not breaking,
- we need more react 18 tests (like `tests/transition.test.tsx`) before refactoring,
- we are still not sure if this is fixable, and
- it's worth investigating.

So, "needs more tests, needs investigation".

**Comment by dai-shi** [comment]
I wouldn't label it a bug.

My future refactor (#1244, or something) might change the behavior. Don't know yet.

**Comment by dai-shi** [comment]
This is fixed in v2 API (pre-released since v1.10.0).
https://codesandbox.io/s/jotai-createstore-derived-sub-forked-r29ug6?file=/src/App.tsx
